### PR TITLE
Vectorize torch.exp2 on CPU and add complex support

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92632
* #92116
* __->__ #92115

I see an 11x speedup in `exp2` on CPU from this.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by sanchitintel** [comment]
Hi @peterbell10 @mingfeima @jgong5,

do we have some sort of a list of ops which are still not vectorized?

Thanks!

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 1]
```diff
@@ -379,6 +379,9 @@ struct Vectorized {
   Vectorized<T> exp() const {
     return map(std::exp);
   }
+  Vectorized<T> exp2() const {
+    return map([](T x) -> T { return exp2_impl(x); });
```
Comment (lezcano) [inline aten/src/ATen/cpu/vec/vec_base.h:383]: Why not simply
```suggestion
    return map(exp2_impl);
```

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_double_vsx.h [hunk 2]
```diff
@@ -240,6 +240,9 @@ class Vectorized<double> {
   Vectorized<double> C10_ALWAYS_INLINE exp() const {
      return {Sleef_expd2_u10vsx(_vec0), Sleef_expd2_u10vsx(_vec1)};
   }
+  Vectorized<double> C10_ALWAYS_INLINE exp2() const {
+    return {Sleef_exp2d2_u10vsx(_vec0), Sleef_exp2d2_u10vsx(_vec1)};
```
Comment (lezcano) [inline aten/src/ATen/cpu/vec/vec256/vsx/vec256_double_vsx.h:244]: The returns around this one have a 3 space indentation. Weird.

File: aten/src/ATen/cpu/vec/vec256/vec256_complex_float.h [hunk 3]
```diff
@@ -268,6 +268,12 @@ template <> class Vectorized<c10::complex<float>> {
...
     return _mm256_mul_ps(exp, cos_sin);
   }
+  Vectorized<c10::complex<float>> exp2() const {
+    // Use identity 2**x = exp(log(2) * x)
+    const __m256 ln_2 = _mm256_set1_ps(c10::ln_2<float>);
+    Vectorized<c10::complex<float>> scaled_values = _mm256_mul_ps(values, ln_2);
+    return scaled_values.exp();
+  }
   Vectorized<c10::complex<float>> expm1() const {
     AT_ERROR("not supported for complex numbers");
```
Comment (lezcano) [inline aten/src/ATen/cpu/vec/vec256/vec256_complex_float.h:278]: Not for this PR, but ugh.

**Review by lezcano — APPROVED** [review]
Nice!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**